src/views/Users/ViewUser/index.jsx

```diff
@@ -0,0 +1,470 @@
+        return entry;
+      }
+
+      const result = entry;
+
+      result.options.actions[index] = value;
+
+      return result;
+    };
+
+    return permission.metadata.isAdditional
+      ? setAdditionalPermissions(additionalPermissions.map(setAction))
+      : setPermissions(permissions.map(setAction));
+  };
+
+  const handleActionDelete = (permission, index) => {
+    const removeAction = entry => {
+      if (entry.name !== permission.name) {
+        return entry;
+      }
+
+      const result = entry;
+
+      result.options.actions.splice(index, 1);
+
+      return result;
+    };
+
+    return permission.metadata.isAdditional
+      ? setAdditionalPermissions(additionalPermissions.map(removeAction))
+      : setPermissions(permissions.map(removeAction));
+  };
+
+  const handlePermissionAdd = () => {
+    setAdditionalPermissions(
+      additionalPermissions.concat([getEmptyPermission()])
+    );
+  };
+
+  const handlePermissionNameChange = (permission, index, value) => {
+    const setName = additionalPermissions.map(entry => {
+      if (entry.name !== permission.name) {
+        return entry;
+      }
+
+      const result = entry;
+
+      result.name = value;
+
+      return result;
+    });
+
+    setAdditionalPermissions(setName);
+  };
+
+  const handleUserSave = () => {};
+  const handleUserDelete = () => {};
+  const renderRole = (role, index) => (
+    <Grid container spacing={2} key={index}>
+      <Grid item xs>
+        <TextField
+          onChange={e => handleRoleNameChange(role, index, e.target.value)}
+          value={role.name}
+        />
+      </Grid>
+      <Grid item xs>
+        <IconButton onClick={() => handleRoleDelete(role, index)}>
+          <DeleteIcon />
+        </IconButton>
+      </Grid>
+    </Grid>
+  );
+  const renderPermission = permission =>
+    zip(
+      [permission.name],
+      ((permission.options && permission.options.products) || []).concat([
+        'add',
+      ]),
+      ((permission.options && permission.options.actions) || []).concat(['add'])
+    ).map((row, index) => (
+      // TODO: need to replace key with something that's not the index, but also won't
+      // change when the data changes.
+      <Grid container spacing={2} key={index} className={classes.gridWithIcon}>
+        <Grid item xs>
+          {row[0] !== undefined && (
+            <TextField
+              required
+              value={row[0]}
+              className={classes.fullWidth}
+              disabled={permission.metadata.isAdditional ? false : !isNewUser}
+              onChange={e =>
+                handlePermissionNameChange(permission, index, e.target.value)
+              }
+            />
+          )}
+        </Grid>
+        {row[1] === undefined && <Grid item xs />}
+        {row[1] !== undefined && row[1] !== 'add' && (
+          <Fragment>
+            <Grid item xs>
+              <TextField
+                value={row[1]}
+                style={{ width: '85%' }}
```

Comment:
For readability, I find separating styling from the jsx helps. We could refer to a class to attach styles.

Reply:
Fixed this with a nested Grid instead.